src/charm.py

```diff
@@ -2,134 +2,106 @@
-        "content-type",
-        "grpc-timeout",
-        "keep-alive",
-        "user-agent",
-        "x-accept-content-transfer-encoding",
-        "x-accept-response-streaming",
-        "x-grpc-web",
-        "x-user-agent",
-        "custom-header-1",
-    ]
+        self.charm_reconciler = CharmReconciler(self)
 
-    virtual_host = api.route.VirtualHost(
-        name="local_service",
-        domains=["*"],
-        routes=[
-            api.route.Route(
-                match=api.route.RouteMatch(prefix="/"),
-                route=api.route.RouteAction(
-                    cluster=cluster,
-                    max_grpc_timeout=timedelta(seconds=60),
-                ),
+        self.leadership_gate = self.charm_reconciler.add(
+            component=LeadershipGateComponent(
+                charm=self,
+                name="leadership-gate",
             )
-        ],
-        cors=api.route.CorsPolicy(
-            allow_origin=["*"],
-            allow_methods=",".join(allowed_methods),
-            allow_headers=",".join(allowed_headers),
-            max_age="1728000",
-            expose_headers="grpc-status,grpc-message,custom-header-1",
-        ),
-    )
+        )
 
-    filter_listener = api.listener.Filter(
-        name="envoy.http_connection_manager",
-        config=hcm.HttpConnectionManager(
-            codec_type=hcm.HttpConnectionManagerCodecType.AUTO,
-            stat_prefix="ingress_http",
-            route_config=api.RouteConfiguration(
-                name="local_route",
-                virtual_hosts=[virtual_host],
+        # TODO before ckf 1.9: Change this from SDI to the service-info lib
+        #  https://github.com/canonical/envoy-operator/issues/76
+        self.grpc = self.charm_reconciler.add(
+            component=SdiRelationDataReceiverComponent(
+                charm=self,
+                name="relation:grpc",
+                relation_name="grpc",
             ),
-            http_filters=[
-                hcm.HttpFilter(name="envoy.grpc_web"),
-                hcm.HttpFilter(name="envoy.cors"),
-                hcm.HttpFilter(name="envoy.router"),
-            ],
-        ),
-    )
-    return api.Listener(
-        name="listener_0",
-        address=api.core.Address(
-            socket_address=api.core.SocketAddress(
-                address="0.0.0.0",
-                port_value=port,
-            )
-        ),
-        filter_chains=[api.listener.FilterChain(filters=[filter_listener])],
-    )
-
-
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
+            depends_on=[self.leadership_gate],
+        )
 
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
+        # Should this Component block the charm if it is not available?
+        # It is a requirement of the Kubeflow bundle, not the envoy charm.
+        # Envoy doesn't  NEED this relation, but Charmed Kubeflow using this for
```

Comment:
Yeah I see your point.  It felt strange introducing `metadata-envoy` here since it isn't mentioned anywhere else (at least, until we rename the charm), so I tried to reword the comment a bit instead.  wdyt?

Reply:
lgtm